tools/tpm2_startauthsession.c

```diff
@@ -0,0 +1,138 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <errno.h>
+#include <inttypes.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+#include <ctype.h>
+
+#include <sapi/tpm20.h>
+
+#include "files.h"
+#include "log.h"
+#include "tpm2_alg_util.h"
+#include "tpm2_options.h"
+#include "tpm2_session.h"
+#include "tpm2_tool.h"
+#include "tpm2_util.h"
+
+typedef struct tpm2_startauthsession_ctx tpm2_startauthsession_ctx;
+struct tpm2_startauthsession_ctx {
+    struct {
+        TPM2_SE type;
+        TPMI_ALG_HASH halg;
+    } session;
+    struct {
+        const char *path;
+    } output;
+};
+
+static tpm2_startauthsession_ctx ctx = {
+    .session = {
+        .type = TPM2_SE_TRIAL,
+        .halg = TPM2_ALG_SHA256
+    }
+};
+
+static bool on_option(char key, char *value) {
+
+    switch (key) {
+    case 'a':
+        ctx.session.type = TPM2_SE_POLICY;
+        break;
+    case 'g':
+        ctx.session.halg = tpm2_alg_util_from_optarg(value);
+        if(ctx.session.halg == TPM2_ALG_ERROR) {
+            LOG_ERR("Invalid choice for policy digest hash algorithm");
+            return false;
+        }
+        break;
+    case 'S':
+        ctx.output.path = value;
+        break;
+    }
+
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    static struct option topts[] = {
+        { "auth-policy-session", no_argument,       NULL, 'a'},
+        { "policy-digest-alg",   required_argument, NULL, 'g'},
+        { "session",             required_argument, NULL, 'S'},
+    };
+
+    *opts = tpm2_options_new("ag:S:", ARRAY_LEN(topts), topts, on_option,
+                             NULL, false);
+
+    return *opts != NULL;
+}
+
+int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
+
+    UNUSED(flags);
+
+    int rc = 1;
+    tpm2_session_data *session_data = tpm2_session_data_new(ctx.session.type);
+    if (!session_data) {
+        LOG_ERR("oom");
+        return rc;
+    }
+
+    tpm2_session_set_authhash(session_data, ctx.session.halg);
+
+    tpm2_session *s = tpm2_session_new(sapi_context,
+            session_data);
+    if (!s) {
```

Comment:
You have to free tpm2_session_data *session_data before returning.

Reply:
Nope. tpm2_session_new() takes ownership. So when it fails it should free session_data. I did it specifically to make this code easier to write. Nothing should fail from tpm2_session_data_new() to when you call this, and you can avoid dealing with the session_data pointer once you call this.

```
tpm2_session *tpm2_session_new(TSS2_SYS_CONTEXT *sapi_context,
        tpm2_session_data *data) {

    tpm2_session *session = calloc(1, sizeof(tpm2_session));
    if (!session) {
        free(data);
        LOG_ERR("oom");
        return NULL;
    }

    session->input = data;

    session->internal.nonceNewer.size = session->input->nonce_caller.size;

    if (!sapi_context) {
        return session;
    }

    bool result = start_auth_session(sapi_context, session);
    if (!result) {
        tpm2_session_free(&session);
        return NULL;
    }

    return session;
}
```